Feature/person overview

azurebrd opened the pull request:
Linting with 'lein eastwood'
There are 2 warnings from gene_test.clj and 1 warning from test_routing.clj, which I didn't modify, so I don't know whether warnings are okay to ignore.
There are 2 warnings from publications.clj about def-in-def. I don't know why it's bad, nor alternatively, how to get around doing group-by without defining results into a new variable.

azurebrd committed:
Initializing files for person overview

azurebrd committed:
Stubs for overview and street address

azurebrd committed:
Syntax errors

azurebrd committed:
Different syntax error

azurebrd committed:
Working except for Old Address

azurebrd committed:
Working version of Person Overview widget with Institution field commented out awaiting WS258 data fix.

azurebrd committed:
Working Tracking widget.  Added new date format 4.

azurebrd committed:
Working Laboratory widget.

azurebrd committed:
Working Publications widgets.

azurebrd committed:
Linting with lein eastwood.
There are 2 warnings from gene_test.clj and 1 warning from test_routing.clj, which I didn't modify, so I don't know whether warnings are okay to ignore.
There are 2 warnings from publications.clj about def-in-def which I don't know why it's bad, or how to get around doing group-by without defining results into a new variable.

azurebrd committed:
Removed commented line

sibyl229 commented:
Hi @azurebrd def-in-def is probably not what you want, because `def` is global to the namespace, maybe what you need is `let`? 

I haven't tried eastwood lately to know if the other warnings are okay.

I think the code is nice. I'm happy with everything else. Just some indentations are off, can you try have your text editor reindent everything?

azurebrd commented:
Thanks @sibyl229 ! I'll try out let tomorrow, and I remember now Adam telling me about vim-fireplace for the indenting.

mgrbyte commented on src/rest_api/classes/person/widgets/publications.clj:
```diff
@@ -0,0 +1,71 @@
+
+(defn publications [person]
+   (let [db (d/entity-db person)
+         dataArrayTuples
```
Suggest renaming `dataArrrayTuples` to just `data` or `datoms`

mgrbyte commented on src/rest_api/classes/person/widgets/publications.clj:
```diff
@@ -0,0 +1,71 @@
+                       :brief_citation (:paper/brief-citation paper)
+                     ))))
+              (seq))]
+     (def data (group-by :year dataArrayTuples))
```
You can eliminate the `eastwood` warnings by removing this line (32),
and adding `((group-by :year))` as the final step in the `->>` pipeline, just below `(seq)` above.

mgrbyte commented on src/rest_api/classes/person/widgets/publications.clj:
```diff
@@ -0,0 +1,71 @@
+                       :brief_citation (:paper/brief-citation paper)
+                     ))))
+              (seq))]
+     (def data (group-by :year dataArrayTuples))
```
Same as above w/regard to naming and eliminating `lein eastwood` warnings.

mgrbyte reviewed (commented):
Generally, logic looks good.
Style-wise, we're trying to keep to "standard" clojure conventions:
  naming: var-names-like-this
  spacing: 2 space indent
  braces brackets wrap up (never on their own line)

mgrbyte commented on src/rest_api/classes/person/widgets/laboratory.clj:
```diff
@@ -0,0 +1,122 @@
+   "allele designation of the affiliated laboratory"}))
+;; note that the perl catalyst interface does not return proper data for "rep", it gives a single value while the data could have multiple persons, e.g. WBPerson8705
+
+(defn gene-classes [person]
```
Could this be re-worked so just one datomic query is made?
Currently: 1 + number-of-labs-per-person queries.

Something like:
```clojure
(d/q '[:find [?geneclass ...]
         :where
         [?lab :laboratory/registered-lab-members ?person]
         [?geneclass :gene-class/designating-laboratory ?lab]]
        db (:db/id person))
```

azurebrd committed:
Fixed def-in-def with Matt's suggestion of (group-by :year) following the (seq)

azurebrd committed:
Laboratory xref to Gene_class many to one, Gene_class xref to Laboratory uniquely.
Querying for the Gene_classes, then following with a query to the lab gives a single lab.
This is not necessarily helpful, but doesn't hurt.

azurebrd commented:
Thanks @mgrbyte !  

The (seq)(group-by :year) worked great.

For reducing the number of gene class queries, it makes sense to query for all the gene classes of a given person, and follow them to the gene class's laboratory, but to get the :lab :label/:id is there a way to follow the gene_class without making a new query into each gene_class's laboratory ?  Something like this doesn't work

                                (let [geneclass (d/entity db oid)]
                                  (pace-utils/vmap
                                   :lab
                                      {:taxonomy "all"
                                       :class "laboratory"
                                       :label (:gene-class/designating-laboratory geneclass)

because it's the :ref entity instead of the :laboratory/id

In ACeDB gene_class xrefs many-to-one to laboratory, and laboratory xrefs uniquely to gene_class, so I've switched the order of the querying, gene_class first then laboratory within that, but that doesn't reduce the number of queries, just sort of returns the data in a more pair-like way.

If you meant that we can query for a person's laboratory and gene_class separately and aggregate them, a person can have multiple labs, and each gene_class has to be associated to its own laboratory.  If that weren't the case and we could aggregate the data off of separate queries, I wouldn't know how to write the syntax, but it'd be nice to know that it would be okay to do things that way.  I've been following the format of existing queries to generate the data, but there's been times that I've wanted to query things separately and aggregate afterward.

mgrbyte commented on src/rest_api/classes/person/widgets/publications.clj:
```diff
@@ -57,8 +56,7 @@
                      ))))
-              (seq))]
-     (def data (group-by :year dataArrayTuples))
+              (seq)(group-by :year))]
```
Nearly! 
Think you want:
```
...
(group-by :year)
(seq))
````

adamjohnwright commented on src/rest_api/classes/person/widgets/laboratory.clj:
```diff
@@ -0,0 +1,125 @@
+                                   :class "person"
+                                   :label (:person/standard-name rep)
+                                   :id (:person/id rep)}) ))
+                       )
```
You should pull these braces up

adamjohnwright commented on src/rest_api/classes/person/widgets/laboratory.clj:
```diff
@@ -0,0 +1,125 @@
+                                   :label (:person/standard-name rep)
+                                   :id (:person/id rep)}) ))
+                       )
+                     )))
```
You should pull these braces up

adamjohnwright commented on src/rest_api/classes/person/widgets/laboratory.clj:
```diff
@@ -0,0 +1,125 @@
+                                               :class "gene_class"
+                                               :label (:gene-class/id geneclass)
+                                               :id (:gene-class/id geneclass)})
+                        ) )))
```
You should pull these braces up

adamjohnwright commented on src/rest_api/classes/person/widgets/laboratory.clj:
```diff
@@ -0,0 +1,125 @@
+                              "no representative")
+                         )
+                       ))
+                     )))
```
You should pull these braces up

adamjohnwright commented on src/rest_api/classes/person/widgets/overview.clj:
```diff
@@ -0,0 +1,122 @@
+                       :other-phone (:address/other-phone old-address)
+                       :fax (:address/fax old-address)
+                       :web-page (:address/web-page old-address)
+                      )
```
You should pull these braces up

adamjohnwright commented on src/rest_api/classes/person/widgets/publications.clj:
```diff
@@ -0,0 +1,69 @@
+(def widget
+  { :publications             publications
+    :meeting_abstracts        meeting-abstracts
+  })
```
You should pull these braces up

adamjohnwright commented on src/rest_api/classes/person/widgets/tracking.clj:
```diff
@@ -0,0 +1,34 @@
+  (:require
+   [datomic.api :as d]
+   [rest-api.formatters.date :as date]))
+
```
Should remove extra lines like this one

azurebrd committed:
Fixed indenting with vim-fireplace

azurebrd commented:
@a8wright Went through formatting with vim-fireplace and cleaned it up.

adamjohnwright commented:
you will need to do a "git merge develop" and then fix the merge conflicts. Then do a merge commit. and push. I think it will be ready to pull at that point!  Great work!!

azurebrd commented:
@a8wright Thanks Adam !  I think it's done.  I ended up in the development branch, where I merged, fixed, and pushed it.  I then came back to this branch, merged in develop, saw the changes and it all runs well, so I think it's good.  I think I just didn't 'git pull' before merging develop, which is why it didn't have any changes in develop to merge.  Sorry about that.

adamjohnwright commented:
@azurbrd No problem. I use the gitflow tool and it doesn't pull from develop for me. I thought the "git hf update" would do it but no such luck lately. I am glad you are able to get it merged.

azurebrd commented:
Thanks !
